bin/cloud_builder.sh

```diff
@@ -0,0 +1,135 @@
+
+OUTPUT_PATH=""
+# The default for PROXY_PATH (which indicates where the proxy path is located
+# relative to the istio repo) is based on repo manifest that places istio at:
+# go/src/istio.io/istio
+# and proxy at:
+# src/proxy
+PROXY_PATH="../../../../src/proxy"
+TAG_NAME="0.0.0"
+
+function usage() {
+  echo "$0
+    -o        path to store build artifacts
+    -p        path to proxy repo (relative to istio repo, defaults to ${PROXY_PATH} ) 
+    -t <tag>  tag to use (optional, defaults to ${TAG_NAME} )"
+  exit 1
+}
+
+while getopts o:p:t: arg ; do
+  case "${arg}" in
+    o) OUTPUT_PATH="${OPTARG}";;
+    p) PROXY_PATH="${PROXY_PATH}";;
+    t) TAG_NAME="${OPTARG}";;
+    *) usage;;
+  esac
+done
+
+[[ -z "${OUTPUT_PATH}" ]] && usage
+[[ -z "${PROXY_PATH}" ]] && usage
+
+# switch to the root of the istio repo
+ROOT="$(cd "$(dirname "${BASH_SOURCE[0]}")/.." && pwd)"
+cd $ROOT
+
+export GOPATH="$(cd "$ROOT/../../.." && pwd)":${ROOT}/vendor
+echo gopath is $GOPATH
+
+if [ ! -d "${PROXY_PATH}" ]; then
+  echo "proxy dir not detected at ${PROXY_PATH}"
+  usage
+fi
+
+# Proxy has some specific requirements for Bazel's
+# config (plus it's nicely places bazel in batch
+# mode) so this component gets built first.
+
+pushd "${PROXY_PATH}"
+
+# Use this file for Cloud Builder specific settings.
+# This file sets RAM sizes and also specifies batch
+# mode that should shutdown bazel after each call.
+echo 'Setting bazel.rc'
+cp tools/bazel.rc.cloudbuilder "${HOME}/.bazelrc"
+
+./script/push-debian.sh -c opt -v "${TAG_NAME}" -o "${OUTPUT_PATH}"
+popd
+
+# Pilot likes to check if the source tree is 'clean'
+# when it queries for version/source informatin.  Some
+# other components like littering the tree so it's better
+# to build pilot sooner than later.
+
+# Pilot build expects this file to exist.  The usual
+# approach of adding a symlink to the user's config file
+# doesn't help when the user doesn't have one.
+touch pilot/platform/kube/config
+
+# building //... results in dirtied files:
+# broker/pkg/model/config/mock_store.go
+# broker/pkg/platform/kube/crd/types.go
+# mixer/template/apikey/go_default_library_handler.gen.go
+# mixer/template/apikey/go_default_library_tmpl.pb.go
+# mixer/template/template.gen.go
+bazel build //pilot/...
+
+# bazel_to_go likes to run from dir with WORKSPACE file
+./bin/bazel_to_go.py
+# Remove doubly-vendorized k8s dependencies that confuse go
+rm -rf vendor/k8s.io/*/vendor
+
+# bazel_to_go.py dirties generated_files and lintconfig.json
+# it's easier to ask git to restore files than add
+# an option to bazel_to_go to not touch them
+git checkout generated_files
+git checkout lintconfig.json
+
+pushd pilot
+./bin/upload-istioctl -r -o "${OUTPUT_PATH}"
+# An empty hub skips the tag and push steps.  -h "" provokes unset var error msg so using " "
+./bin/push-docker -h " " -t "${TAG_NAME}" -b -o "${OUTPUT_PATH}"
+./bin/push-debian.sh -c opt -v "${TAG_NAME}" -o "${OUTPUT_PATH}"
```

Comment:
In all cases, we need an option to specify the docker hub. It would also be nice if we can separate building Debian package from pushing it. Some may not have gcs access. In such cases, storing the package in some specific local folder is easier.

Reply:
Currently the only intended output of cloud_builder.sh is the build artifact files that are placed in the caller-specified local filesystem directory (e.g., "-o /output").  Each docker image is stored as a tar file under a "docker/" subdirectory (e.g., "/output/docker/").  GCS/GCR doesn't come into play until Cloud Builder later runs store_artifacts.sh to upload the /output directory to GCS/GCR, so lack of GCS access doesn't affect the ability to run cloud_builder.sh.

The various existing push-docker scripts use a common/shared istio/bin/docker_lib.sh that defaults HUBS to gcr.io/istio-testing/, so the -h " " is used to override this default so no actual push is made to any gcr or docker registry (this is later done by store_artifacts.sh).  Offhand I can think of two options.  The first is to add a -h parameter to cloud_builder.sh that can be used to override the " " default.  This adds convenience for registry uploads w/o having to run another script, though it does go against cloud_builder.sh's intended behavior of only outputting to the local filesystem.  The second option is that I add functionality to store_artifacts.sh so that it supports uploading to 1 or more of GCR, GCS, or docker (currently the script only supports uploading to both GCS & GCR -- this was done partly because the 0.3 istio release process describing builds as only going to a more public locations after the build had been designated as a weekly or monthly build).